File: backend/engine/settlement_engine.py
"""
GridForge Settlement Engine — Python port of src/engine/SettlementEngine.js

Imbalance computation, price selection, and settlement cash flows.
"""

from __future__ import annotations

from .constants import SP_DURATION_H


def compute_imbalance(actual_physical_mw: float, contracted_mw: float, bm_accepted_mw: float = 0) -> float:
    """
    imbalanceMw = actualPhysicalMw − (contractedMw + bmAcceptedMw)
    Positive = Surplus (Paid SSP); Negative = Shortage (Pays SBP)
    """
    return actual_physical_mw - (contracted_mw + (bm_accepted_mw or 0))


def select_imbalance_price(imbalance_mw: float, sbp: float, ssp: float) -> float:
    return sbp if imbalance_mw < 0 else ssp


def compute_imbalance_settlement(
    actual_physical_mw: float,
    contracted_mw: float,
    bm_accepted_mw: float = 0,
    sbp: float = 50,
    ssp: float = 40,
    sp_duration_h: float = SP_DURATION_H,
) -> dict:
    imbalance_mw = compute_imbalance(actual_physical_mw, contracted_mw, bm_accepted_mw)
    price = select_imbalance_price(imbalance_mw, sbp, ssp)
    mwh = imbalance_mw * sp_duration_h
    cash = mwh * price

    return {
        "imbalanceMw": imbalance_mw,
        "price": price,
        "mwh": mwh,
        "cash": cash,
    }


def compute_hub_fee_from_settlements(settlements: list[dict]) -> dict:
    sum_player_imb_cash = sum(s.get("imbCash", 0) for s in settlements)
    return {"sumPlayerImbCash": sum_player_imb_cash, "hubFee": -sum_player_imb_cash}

File: frontend/src/engine/SettlementEngine.js
import { SP_DURATION_H } from "../shared/constants.js";

/**
 * imbalanceMw = actualPhysicalMw − (contractedMw + bmAcceptedMw)
 * Positive = Surplus (Paid SSP); Negative = Shortage (Pays SBP)
 */
export function computeImbalance(actualPhysicalMw, contractedMw, bmAcceptedMw = 0) {
  return actualPhysicalMw - (contractedMw + (bmAcceptedMw || 0));
}

export function selectImbalancePrice(imbalanceMw, sbp, ssp) {
  return imbalanceMw < 0 ? sbp : ssp;
}

export function computeImbalanceSettlement({
  actualPhysicalMw,
  contractedMw,
  bmAcceptedMw = 0,
  sbp,
  ssp,
  spDurationH = SP_DURATION_H,
}) {
  const imbalanceMw = computeImbalance(actualPhysicalMw, contractedMw, bmAcceptedMw);
  const price = selectImbalancePrice(imbalanceMw, sbp, ssp);
  const mwh = imbalanceMw * spDurationH;
  const cash = mwh * price;

  return {
    imbalanceMw,
    price,
    mwh,
    cash,
  };
}

export function computeHubFeeFromSettlements(settlements) {
  const sumPlayerImbCash = settlements.reduce((acc, s) => acc + (s.imbCash || 0), 0);
  return { sumPlayerImbCash, hubFee: -sumPlayerImbCash };
}

